textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/join/Join.java

```diff
@@ -42,14 +88,160 @@ public void open() throws Exception {
+			return null;
+		}
+
 		ITuple nextTuple = null;
+
+		do {
+			if(getOuterOperatorNextTuple == true) {
+				if((outerTuple = outerOperator.getNextTuple()) != null) {
+					getOuterOperatorNextTuple = false;
+				} else {
+					return null;
+				}
+			}
+
+			if (innerOperatorCursor <= innerTupleList.size() - 1) {
+				innerTuple = innerTupleList.get(innerOperatorCursor);
+				innerOperatorCursor++;
+				if(innerOperatorCursor == innerTupleList.size()) {
+					innerOperatorCursor = 0;
+					getOuterOperatorNextTuple = true;
+				}
+			}
+
+			nextTuple = processTuples(outerTuple, innerTuple, joinPredicate);
+		} while(nextTuple == null);
+
 		return nextTuple;
 	}
 
 	@Override
 	public void close() throws Exception {
-		// TODO Auto-generated method stub
-		
+		try {
+			outerOperator.close();
+			// innerOperator.close(); already called in open()
+
+		} catch(Exception e) {
+			e.printStackTrace();
+			throw new DataFlowException(e.getMessage(), e);
+		}
+		// Clear the inner tuple list from memory on close.
+		innerTupleList.clear();
+	}
+
+	// Used to compare IDs of the tuples.
+	private boolean compareId(ITuple outerTuple, ITuple innerTuple) {
+		// TODO(Flavio): what if the join predicate has an invalid id attribute?
+		// First check if the field in question exits by using try catch.
+		// Will throw an exception if it doesn't exist. This leads to return false.
+		// Then check if both the fields are of type IntegerField.
+		// --> This is the bare minimum thing that can be done to to verify valid 
+		// id attribute (as of now) (probably it is better to add a field called ID).
+		String fieldName = joinPredicate.getidAttribute().getFieldName();
+		try {
+			if(outerTuple.getField(fieldName).getClass().equals(IntegerField.class)&&
+					innerTuple.getField(fieldName).getClass().equals(IntegerField.class)) {
+				if(outerTuple.getField(fieldName).getValue()==
+						innerTuple.getField(fieldName).getValue()) {
+					return true;
+				}
+			}
+		} catch(Exception e) {
+			;
+		}
+		return false;
+	}
+
+	// Process the tuples to get a tuple with join result if predicate is satisfied.
+	private ITuple processTuples(ITuple outerTuple, ITuple innerTuple, JoinPredicate joinPredicate) throws Exception {
+		ITuple nextTuple = null;
+		List<Span> newJoinSpanList = new ArrayList<>();
+
+		if(!compareId(outerTuple, innerTuple)) {
+			return null;
+		}
+		//TODO(Flavio): getIndex will always return an int, if SchemaConstants.SPAN_LIST doesn't 
+		// exists a NullPointerException will be thrown, it will never return null and the
+		// following comparison will be useless
+		// If either/both tuples have no span information, return null.
+		Integer indexOfInnerSpanList = null;
+		Integer indexOfOuterSpanList = null;
+		try {
+			indexOfInnerSpanList = innerTuple.getSchema().getIndex(SchemaConstants.SPAN_LIST);
+			indexOfOuterSpanList = outerTuple.getSchema().getIndex(SchemaConstants.SPAN_LIST);
+		} catch(Exception e) {
```

Comment:
Why do we get the index of spanList here?
If we want the spanList, we can just do `tuple.getField(SchemaConstants.SPAN_LIST)`